xchange-stream-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingAdapters.java

```diff
@@ -19,58 +14,93 @@
 import org.knowm.xchange.kraken.dto.trade.KrakenType;
 import org.knowm.xchange.utils.DateUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
-/** Kraken streaming adapters */
-public class KrakenStreamingAdapters {
-
-  static final String ASK_SNAPSHOT = "as";
-  static final String ASK_UPDATE = "a";
+import java.math.BigDecimal;
+import java.time.Instant;
+import java.util.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
 
-  static final String BID_SNAPSHOT = "bs";
-  static final String BID_UPDATE = "b";
+import static info.bitrich.xchangestream.kraken.KrakenStreamingChecksum.createCrcChecksum;
 
-  private static final List<String> BID_KEYS = Lists.newArrayList(BID_SNAPSHOT, BID_UPDATE);
-  private static final List<String> ASK_KEYS = Lists.newArrayList(ASK_SNAPSHOT, ASK_UPDATE);
+/**
+ * Kraken streaming adapters
+ */
+public class KrakenStreamingAdapters {
+    private static final Logger LOG = LoggerFactory.getLogger(KrakenStreamingAdapters.class);
+
+    static final String ASK_SNAPSHOT = "as";
+    static final String ASK_UPDATE = "a";
+
+    static final String BID_SNAPSHOT = "bs";
+    static final String BID_UPDATE = "b";
+
+    static final String CHECKSUM = "c";
+
+    private static void updateInBook(int depth, Instrument instrument, Order.OrderType orderType, JsonNode currentNode, String key, TreeSet<LimitOrder> target) {
+        adaptLimitOrders(instrument, orderType, currentNode.get(key)).forEachRemaining(limitOrder -> {
+            target.removeIf(it -> it.getLimitPrice().compareTo(limitOrder.getLimitPrice()) == 0);
+            if (limitOrder.getOriginalAmount().compareTo(BigDecimal.ZERO) != 0) {
+                target.add(limitOrder);
+            }
+        });
+        while (target.size() > depth) {
+            LimitOrder last = target.last();
+            target.remove(last);
+        }
+    }
 
-  public static OrderBook adaptOrderbookMessage(
-      OrderBook orderBook, Instrument instrument, ArrayNode arrayNode) {
-    Streams.stream(arrayNode.elements())
-        .filter(JsonNode::isObject)
-        .forEach(
-            currentNode -> {
-              for (String key : BID_KEYS) {
-                if (currentNode.has(key)) {
-                  adaptLimitOrders(instrument, Order.OrderType.BID, currentNode.get(key))
-                      .forEach(orderBook::update);
+    public static OrderBook adaptOrderbookMessage(int depth, AtomicBoolean awaitingSnapshot, TreeSet<LimitOrder> bids, TreeSet<LimitOrder> asks, Instrument instrument, ArrayNode arrayNode) {
+        final AtomicLong expectedChecksum = new AtomicLong(0);
+        final AtomicReference<Date> lastTime = new AtomicReference<>(Date.from(Instant.EPOCH));
+        arrayNode.elements().forEachRemaining(currentNode -> {
+            if (awaitingSnapshot.get()) {
```

Comment:
do we expect a scenario where we get updates before the snapshot? why do we need this await

Reply:
that logic is in place to ensure it doesnt keep applying snapshots when the state is cleared, that would result in a subscribe loop.

Reply:
This has been simplified a bit by just using empty bid/ask maps as 'awaiting snapshot' indicator.